assemble/Particle_Diagnostics.F90

```diff
@@ -0,0 +1,1420 @@
+!    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
+!    USA
+
+#include "fdebug.h"
+
+module particle_diagnostics
+
+  use fldebug
+  use global_parameters, only : OPTION_PATH_LEN, FIELD_NAME_LEN
+  use futils, only: int2str
+  use spud
+  use parallel_tools, only: getnprocs, allsum, getprocno
+  use elements, only: eval_shape
+  use fields_base, only: ele_val, ele_loc, node_val
+  use parallel_fields, only: node_owned
+  use fields_calculations, only: dot_product
+  use fields
+  use profiler
+  use state_module
+  use halos
+  use detector_data_types
+  use pickers
+  use field_options
+  use detector_tools, only: temp_insert, insert, allocate, deallocate, temp_deallocate, &
+       & remove, temp_remove
+  use particles, only : get_particle_arrays, initialise_constant_particle_attributes, &
+       & update_particle_attributes_and_fields, get_particle_arrays
+  use multimaterial_module, only: calculate_sum_material_volume_fractions
+
+  implicit none
+
+  private
+
+  public :: initialise_particle_diagnostics, update_particle_diagnostics, &
+       & calculate_diagnostics_from_particles, calculate_field_from_particles, &
+       & initialise_constant_particle_diagnostics, &
+       & initialise_particle_diagnostic_fields_post_adapt, particle_cv_check
+
+  contains
+
+  subroutine initialise_constant_particle_diagnostics(state)
+    !subroutine to initialise constant particle attributes and
+    !MVF fields if 'from_particles'
+
+    use particles, only: particle_lists
+    type(state_type), dimension(:), intent(inout) :: state
+
+    character(len = OPTION_PATH_LEN) :: group_path, subgroup_path
+    type(scalar_field), pointer :: s_field
+    integer :: i, j
+    integer :: particle_groups, list_counter, particle_materials
+    integer, dimension(:), allocatable :: particle_arrays
+
+    type(detector_type), pointer :: particle
+
+    ewrite(2,*) "In initialise_constant_particle_diagnostics"
+
+    !Check if there are particles
+    particle_groups = option_count("/particles/particle_group")
+
+    if (particle_groups==0) return
+
+    !Set up particle_lists
+    allocate(particle_arrays(particle_groups))
+    particle_arrays(:) = 0
+    do i = 1,particle_groups
+       particle_arrays(i) = option_count("/particles/particle_group["//int2str(i-1)//"]/particle_subgroup")
+    end do
+
+    !Initialise constant particle attributes
+    list_counter=1
+    do i = 1,particle_groups
+       group_path = "/particles/particle_group["//int2str(i-1)//"]"
+       do j = 1,particle_arrays(i)
+          particle => particle_lists(list_counter)%first
+          subgroup_path = trim(group_path) // "/particle_subgroup["//int2str(j-1)//"]"
+          if (option_count(trim(subgroup_path) // "/attributes/scalar_attribute/constant") + &
+               & option_count(trim(subgroup_path) // "/attributes/vector_attribute/constant") + &
+               & option_count(trim(subgroup_path) // "/attributes/tensor_attribute/constant").gt.0) then
```

Comment:
Nitpicking on style: could you just use `>`, `<`, `>=`, `<=` and `==` rather than `.gt.', etc.